### ISSUE TITLE
Suggestion: Tag or standard method to filter lines to relay clients

### ISSUE THREAD

**Issue by khaytsus** [weechat/weechat#1619]
There are things that I have in my local weechat client I want to see, such as script prints, part/joins, etc, that I do not need to see on any relay clients, such as a mobile device.  I know that clients can generally hide part/join themselves, but I see this useful for scripts that might have verbose output that are useful locally but not mobile.

I have been thinking about the idea that weechat itself does not send lines tagged a certain way to relay clients, possibly saving some data on the client end but would require new functionality in weechat and ways to set it, toggle it, etc.  This could also be fairly simple and just use some agreed-upon tag names (example: "no_relay" or "no_relay_send") that weechat itself does not relay at all, no need for any real configuration.

A simpler option would be simply relay clients offer the ability to hide arbitrary tags a user defines or an agreed-upon tag, something like relay_hide.  In this case I would likely need to open bugs against clients themselves to make the suggestions there, since weechat tags are already an arbitrary thing.

**Comment by flashcode** [comment]
Hi,

First I moved the issue to WeeChat repository, since it's something with WeeChat, not weechat-relay which is a lib just for encoding/decoding relay messages, so at a raw level (and it's work in progress, not used yet by WeeChat itself).

Relay clients already have everything they need to filter lines: tags, and even a flag `displayed` which means the line is displayed in WeeChat (or filtered with a filter), see doc here: https://weechat.org/files/doc/devel/weechat_relay_protocol.en.html#message_buffer_line_added

So for me it's something you could ask to relay client developers, I'm not sure there's something to implement on WeeChat side.

**Comment by khaytsus** [comment]
Yep, requesting that clients offer some ability to filter either a fixed tag or a user-input tag is certainly a simple solution.  Saving a little mobile data is nice but I'll leave that up to you if you think that's valuable to the project overall.

**Comment by flashcode** [comment]
OK, I see your point: you would like to save data exchange by not sending lines that wouldn't be displayed in the relay client at all.
So if you do that, you agree that the relay client will never be able to show these lines (like WeeChat can do with filtered lines, which are just hidden but present in the buffer).

So this could be implemented by adding for example a new option in `sync` command that would tell to WeeChat to not send hidden lines for examples (filtered by WeeChat), or on another condition.

**Comment by khaytsus** [comment]
Yep, I think if the tag is documented clearly and is a new tag, it seems safe that nobody would accidentally use it or already be using it and wind up filtering unexpectedly.  I wouldn't want to necessarily remove ALL filtered lines, or maybe that's a useful option as well?  I'm a pretty new user to weechat, so maybe that's basically the same thing I'm asking for.

**Comment by weechatter** [comment]
I agree with khaytsus, to not send messages with a "specific" tag. this way all clients behave in the same way and do not display those messages, because they do not get the lines at all.

The other way round you have to ask every client author to implement this feature and if one disagree you may have clients
they work different.

My suggestion would be to not send a message with a tag like "no_send_relay" via relay. every script can set such a tag (for example the stalker.pl script, the output from script can be huge!). the output will be print locally, but not in relay

**Comment by trygveaa** [comment]
> My suggestion would be to not send a message with a tag like "no_send_relay" via relay. every script can set such a tag (for example the stalker.pl script, the output from script can be huge!). the output will be print locally, but not in relay

Someone may want to see such lines in the relay client though, so then I guess it should be configurable.

**Comment by weechatter** [comment]
Its nothing different to tags like already existing ones: "no_log" or "no_highlight" trygveaa ;-)

stalker.pl has for example this option to set tags for messages created by the script itself:
/set plugins.var.perl.stalker.tags "no_log,no_highlight"
-> [stalker] Found nicks: nils_2 from host ~nils_2@p5795DFD8.dip0.t-ipconnect.de [no_log,no_highlight]

/set plugins.var.perl.stalker.tags "no_log,no_highlight,no_send_relay"
-> [stalker] Found nicks: nils_2 from host ~nils_2@p5795DFD8.dip0.t-ipconnect.de [no_log,no_highlight,no_send_relay]

This way people can decide if a message will be logged or highlighted or if it should be send to the relay.

Its up to the user if he set the tag or not. i don't want to log the stalker output to my logs, for example. it will
simply blow up the logs, for nothing.

**Comment by trygveaa** [comment]
Hm, yeah, I suppose you're right.

**Comment by flashcode** [comment]
At the beginning, relay was designed to be able to connect with a remote interface, so the goal is to see **exactly** the same things as in WeeChat itself.

Now I understand the need to filter and then see less things than in WeeChat itself, but I'm not sure that WeeChat nor the scripts themselves should care about "relay" plugin by using specific relay tags.

Instead I think sending everything should be the default for all clients (like relay was designed), and relay clients should be able to tell WeeChat/relay what they want or not, maybe with a new command to filter lines.

This could be done in different ways, for example you could filter by:

* regular expression
* tags (existing tags)
* message displayed or not (filtered in WeeChat)
* etc.

**Comment by oakkitten** [comment]
weechat-android supports partial synchronization mode; it would be quite useful if we could subscribe to highlights and private messages from buffers without synchronizing all lines. as i understand, this would require filtering by notify level as well.
